B2g dev force issuer

shane-tomlinson opened the pull request:
@jedp and I are pow-wowing on this.

ozten committed:
Making forceIssuer optional on address_info and cert_key. Extracted withType.

Rewrote patch to work with latest primary-transition-ux patch on b2g tip.

Conflicts:
	resources/static/dialog/views/set_password.ejs

Conflicts:
	resources/static/dialog/js/modules/verify_primary_user.js

ozten committed:
Missed adding new dbutils.js

jedp committed:
Rebased dev and resolved conflicts in these files:

    resources/static/common/js/user.js
    resources/static/dialog/js/misc/helpers.js
    resources/static/dialog/js/modules/authenticate.js
    resources/static/test/cases/common/js/user.js

Conflicts:
	example/rp/index.html
	resources/static/dialog/js/misc/helpers.js
	resources/static/dialog/js/modules/dialog.js

Commit:
i18n-abide requires named variables to be replaced to end with `s`

Commit:
/wsapi/stage_reset and requestPasswordReset no longer take a password.

Conflicts:
	resources/static/common/js/user.js

Commit:
Fix the tests! Mostly updates to function signatures and adding a couple of addresses to the mocks.

Commit:
Set the issuer when the dialog first starts and forget about it.

Remove all the extra forceIssuer code everywhere.

Conflicts:
	resources/static/dialog/js/misc/helpers.js
	resources/static/dialog/js/misc/internal_api.js
	resources/static/dialog/js/misc/state.js
	resources/static/dialog/js/modules/authenticate.js
	resources/static/dialog/js/modules/verify_primary_user.js

Commit:
Remove the *ForceIssuer* functions from storage.js and user.js. Once the issuer is set, it should be forgotten about.

Conflicts:
	resources/static/common/js/user.js

Commit:
Start a stub for fixupIssuer which will scrub the issuer.

Conflicts:
	resources/static/dialog/js/modules/dialog.js

Commit:
Start on the work to get an assertion for the forceIssuer.

Commit:
Add a few basic tests for internal.watch.

When loading up the internal_api, only set user.setIssuer if an issuer is saved for the current domain.

jedp commented on config/local.json:
```diff
@@ -17,7 +17,7 @@
   // enable the measurement of the dom loading
-  "measure_dom_loading": true
+  "measure_dom_loading": true,
+  "forcable_issuers": ["issuer.domain"]
```
I sort of think this should be spelled `forceable_issuers` (not that "forceable" is an English word, either, but it somehow seems better to me.  Interestingly, this becomes one of those words that means two opposite things, like "oversight."  Well, by sound only, I guess.  Forcible vs Forceable)

shane-tomlinson replied:
When I type in "forceable" to Google, it asks "Did you mean forcible?" m-w.com - http://www.merriam-webster.com/dictionary/forceable. Either way, forcable is not the future.

jedp commented on resources/static/common/js/user.js:
```diff
@@ -828,7 +875,7 @@ BrowserID.User = (function() {
      */
     requestTransitionToSecondary: function(email, password, onComplete, onFailure) {
-      User.addressInfo(email, function(info) {
+      User.addressInfo(email, user.forceIssuer, function(info) {
```
Should this be `User.forceIssuer`?

jedp commented on resources/static/dialog/js/modules/authenticate.js:
```diff
@@ -91,6 +90,13 @@ BrowserID.Modules.Authenticate = (function() {
+          enterPasswordState.call(self);
+	} else {
+          createFxAccount.call(self, self.forceIssuer);
+	}
```
Some funky indentation here

jedp commented on resources/static/dialog/views/set_password.ejs:
```diff
@@ -10,13 +10,15 @@
+              <% if (fxaccount) { %>
+                <%= gettext("Your email address is new to us. Please create a password to use with FirefoxOS.") %>
+            <% } else if (transition_no_password) { %>
+                <%- format(gettext("%(idp) no longer allows you to log into Persona with your %(idp) password. Please create a new password to use with your Persona account."), {
```
I think this wants `%(idp)s` instead of `%(idp)` in each case?

shane-tomlinson replied:
That it is, the fix for that hasn't been merged yet. Will do that.

Commit:
fix JSHint errors

Commit:
Fix the success callback for stageAddressForVerification.

Commit:
Cleaning up after rebase from dev. All QUnit tests pass.

Commit:
Fix the broken getAndValidatePassword tests.

The QUnit tests for getAndValidatePassword were updated to use the set_password template whenever the main site signup/signin pages were removed.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -909,6 +936,7 @@ BrowserID.User = (function() {
      * Sync local identities with login.persona.org.  Generally should not need to
      * be called.
      * @method syncEmails
+     * @param {string} - TODO
```
This TODO needs a real entry.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -1284,32 +1327,33 @@ BrowserID.User = (function() {
-          // assertion.
-          User.addressInfo(email, function(info) {
-            if (info.type === "primary") {
+          // TODO what will the type of forceIssuer email addresses be?
```
Is it a secondary?

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -24,7 +24,8 @@ BrowserID.User = (function() {
       userid,
-      auth_status;
+      auth_status,
+      forceIssuer = "default";
```
Rename "forceIssuer" to just "issuer".

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -265,29 +266,35 @@ BrowserID.User = (function() {
-
-    storage.addEmail(email, email_obj);
-    if (onComplete) onComplete(true);
+    // XXX This needs to be looked at to make sure caching does not bite us.
```
Here is an XXX. Will caching bite us?

shane-tomlinson commented on resources/static/common/js/network.js:
```diff
@@ -508,14 +512,15 @@ BrowserID.Network = (function() {
      */
-    certKey: function(email, pubkey, onComplete, onFailure) {
+    certKey: function(email, pubkey, forceIssuer, onComplete, onFailure) {
+      var opts = 'default' === forceIssuer ? {} : {forceIssuer: forceIssuer};
```
For clarity, maybe set a default opts object first, and then add forceIssuer if it is needed.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -1038,6 +1085,7 @@ BrowserID.User = (function() {
      * (is it a primary or a secondary)
      * @method addressInfo
      * @param {string} email - Email address to check.
+     * @param {string} issuer - Force a specific Issuer by specifing a domain. null for default.
```
Has issuer actually been added to the function signature or is this left over?

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -228,6 +240,10 @@ BrowserID.State = (function() {
       }
+      else if(self.newFxAccountEmail) {
+        startAction(false, "doStageUser", info);
+// TODO         startAction(false, "doStageResetPassword", info); ???
```
What is the doStageResetPassword?

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -343,9 +359,10 @@ BrowserID.State = (function() {
 
-      self.email = email;
+      // Maybe use a second global variable so we know which email address was chosen?
+      self.email = user.forceIssuerEmail = email;
```
An accessor function should be used to set the forceIssuerEmail - what is it used for anyways?

shane-tomlinson replied:
@jedp - I cannot find where this is supposed to be used, any ideas?

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -380,6 +397,23 @@ BrowserID.State = (function() {
         }
       }
+      else if (!user.isDefaultIssuer() && !record.cert) {
+        // TODO: Duplicates some of the logic in the authentication action module.
```
If this duplicates some of the logic from the authentication module, should this be moved there?

shane-tomlinson replied:
This also duplicates logic found in pick_email.

These should be collapsed.

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -122,6 +122,12 @@ BrowserID.Modules.Dialog = (function() {
   }
 
+  function fixupIssuer(url) {
+    // XXX do something awesome here.
```
We should indeed so something awesome here - like make sure the issuer is legit.

shane-tomlinson replied:
How are we going to get a list of valid issuers?

shane-tomlinson commented on resources/static/dialog/js/modules/dialog.js:
```diff
@@ -264,6 +270,14 @@ BrowserID.Modules.Dialog = (function() {
+        // forceIsuser is used by the Marketplace to disable primary support
+        // and replace fxos.login.persona.org as the issuer of certs
+        if (paramsFromRP._experimental_forceIssuer) {
+          // TODO check for valid domain
```
This comment can be removed, we are calling fixupIssuer.

shane-tomlinson commented on resources/static/dialog/views/set_password.ejs:
```diff
@@ -10,13 +10,15 @@
-                <%= gettext("Your email address is new to us. Please create a password to use with Persona.") %>
               <% } %>
+            <% } else { %>
+              <%= gettext("Your email address is new to us. Please create a password to use with Persona.") %>
```
I no longer see this message whenever signing up as a new secondary user.

Commit:
Clean up the template logic which figures out which set password text to show

Commit:
Change `forcable` to `forcible`

Commit:
Cleanup.

shane-tomlinson commented:
@jedp, @benadida 

What should the behavior in this sequence be:

Sequence 1
1) User opens dialog from RP that has forceIssuer set
2) User creates account with forceIssuer
3) Site removes forceIssuer from .request call.

OR

Sequence 2
1) User opens dialog from RP that has forceIssuer set
2) User creates account with forceIssuer
3) User visits another RP that does not have forceIssuer set

Should the generated assertion in both of these cases have the "issuer" set if the address was a secondary to begin with?

shane-tomlinson commented on lib/verifier/certassertion.js:
```diff
@@ -105,6 +106,12 @@ function verify(assertion, audience, successCB, errorCB) {
                        HOSTNAME);
+      } else if (issuer === forceIssuer) {
+        if (config.get('forcable_issuers').indexOf(forceIssuer) === -1) {
+          return errorCB("this verifier doesn't won't force issuer for " + forceIssuer);
```
`doesn't won't`?

shane-tomlinson commented on lib/verifier/certassertion.js:
```diff
@@ -105,6 +106,12 @@ function verify(assertion, audience, successCB, errorCB) {
+        if (config.get('forcable_issuers').indexOf(forceIssuer) === -1) {
+          return errorCB("this verifier doesn't won't force issuer for " + forceIssuer);
+        } else {
+          return next(null, publicKey);
```
@jedp - I'm confused by this. If the forceIssuer is the same as the issuer (and an allowed issuer), we call next, but if the forceIssuer is not the same as the issuer, we go on to check delegation? Does that make sense? If an RP requests forceIssuer for a specific domain and then that domain delegates off to another domain, does that break the contract with the RP?

shane-tomlinson commented on lib/wsapi/address_info.js:
```diff
@@ -178,6 +198,10 @@ exports.process = function(req, res) {
       r.state = "unknown";
+    } else if ('default' !== issuer && hasPassword) {
+      r.state = "known";
+    } else if ('default' !== issuer && ! hasPassword) {
```
It took me a few seconds to work through the mental gymnastics of this logic, a comment would be useful.

shane-tomlinson commented on resources/static/common/js/storage.js:
```diff
@@ -48,12 +48,20 @@ BrowserID.Storage = (function() {
-    storage.emails = JSON.stringify(emails);
+  function emailsStorageKey(issuer) {
+    if (issuer && issuer !== "default")
+      return 'forceIssuerEmails';
```
If there are multiple issuers, all issuers will use the same namespace. This should namespace issuers individually.

shane-tomlinson commented on resources/static/common/js/storage.js:
```diff
@@ -48,12 +48,20 @@ BrowserID.Storage = (function() {
 
   function clear() {
     storage.removeItem("emails");
+    storage.removeItem("forceIssuerEmails");
```
Likewise, all forceIssuerEmail namespaces should be removed.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -265,29 +266,35 @@ BrowserID.User = (function() {
+    // cause problems here.
+    User.addressInfo(email, function(info) {
+      var now = new Date();
+      var email_obj = storage.getEmails(forceIssuer)[email] || {
```
Use storage.getEmail here.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -300,13 +307,13 @@ BrowserID.User = (function() {
-      created: new Date()
-    });
+      created: new Date(),
+      verified: options.verified
```
It looks like allowUnverified has crept into here.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -300,13 +307,13 @@ BrowserID.User = (function() {
    * @param {string} options.verified - If the email is 'secondary', is it verified?
    */
-  function persistEmail(options) {
+  function persistEmail(options, issuer) {
```
Rename this persistEmailWithoutKeypair for clarity.

shane-tomlinson commented on resources/static/common/js/user.js:
```diff
@@ -1063,29 +1111,28 @@ BrowserID.User = (function() {
+        // The normalized email is stored in the cache.
+        var normalizedEmail = info.normalizedEmail || email;
+        info.email = normalizedEmail;
+        User.checkEmailIssuer(normalizedEmail, info, function(issuerInfo) {
```
rename issuerInfo back to cleanedInfo, this was a botch on my part.

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -117,6 +117,8 @@ BrowserID.State = (function() {
       self.siteName = info.siteName || info.hostname;
       self.siteTOSPP = !!(info.privacyPolicy && info.termsOfService);
+      var forceIssuer = (!!info.forceIssuer ? info.forceIssuer : 'default');
+      user.setIssuer(forceIssuer);
```
For consistency, this should probably only be set if forceIssuer is not default.

@jedp - How do we handle when an RP changes its forceIssuer? If the communication_iframe has its forceIssuer set, but the dialog does not, there will be a discrepancy between their two views of the world.

shane-tomlinson commented on resources/static/dialog/js/modules/authenticate.js:
```diff
@@ -127,6 +132,18 @@ BrowserID.Modules.Authenticate = (function() {
+        email = getEmail();
+
+    if (email) {
+      self.close("new_fxaccount", { email: email, fxaccount: true }, { email: email });
```
the "fxaccount: true" is implicit in "new_fxaccount".

shane-tomlinson commented on resources/static/dialog/js/modules/authenticate.js:
```diff
@@ -93,6 +91,13 @@ BrowserID.Modules.Authenticate = (function() {
       else if ("primary" === info.type) {
         self.close("primary_user", info, info);
       }
+      else if (!user.isDefaultIssuer()) {
```
This path needs a test.

shane-tomlinson replied:
If non-default issuers completely disable primaries, this check should go above the primary check.

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -380,6 +397,23 @@ BrowserID.State = (function() {
+          // We'll end up in this state again, but we want to see serverInfo.state change
+          user.resetCaches();
+          if (serverInfo.state === "transition_no_password") {
+            var newInfo = _.extend(info, { fxaccount: true });
```
newInfo is created but then never used.

shane-tomlinson replied:
This indicates a test is needed.

shane-tomlinson commented on resources/static/dialog/js/modules/authenticate.js:
```diff
@@ -127,6 +132,18 @@ BrowserID.Modules.Authenticate = (function() {
     }
   }
 
+  function createFxAccount(callback) {
```
Now that there is createFxAccount, maybe createSecondaryUser should be renamed createPersonaAccount for consistency.

Commit:
Cleanup and add tests for the authenticate module.

Commit:
Check to make sure the forced issuer is a hostname. Add unit tests to make sure the check works.

shane-tomlinson commented on resources/static/dialog/js/misc/state.js:
```diff
@@ -380,6 +398,23 @@ BrowserID.State = (function() {
+          user.resetCaches();
+          if (serverInfo.state === "transition_no_password") {
+            var newInfo = _.extend(info, { fxaccount: true });
+            self.newFxAccountEmail = info.email;
```
@jedp, do we really want to create a new account here, or should some FX OS specific email be sent?

Commit:
Allow there to be multiple email issuer namespaces.

Commit:
Fix post-email verification for a user who had a primary email address that was then changed to use forceIssuer.

shane-tomlinson commented on resources/static/dialog/js/modules/pick_email.js:
```diff
@@ -61,6 +62,8 @@ BrowserID.Modules.PickEmail = (function() {
     if (!! record) {
       dialogHelpers.refreshEmailInfo.call(self, email, function (info) {
+        // XXX Why is this here? This is almost a complete duplication of
+        // the logic in state.js, and it should be there.
```
This logic is also found in authenticate.js.

Commit:
Fix JSHint errors.

Commit:
Only set forceIssuer in state.js if a forceIsser is given. Add a test.

Commit:
Remove the unused forceIssuerEmail.
